add slice/select/diagonal_scatter variants as primitive ops

bdhirsh opened the pull request:
The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66101 Test functionalization pass in python
* #64432 Add a Functionalization pass in core
* #65412 add tags for inplace view ops in native_functions.yaml
* **#64430 add slice/select/diagonal_scatter variants as primitive ops**
* #63094 move some codegen utilities into utils.py

Differential Revision: [D31942092](https://our.internmc.facebook.com/intern/diff/D31942092)

bdhirsh committed:
add slice/select/diagonal_scatter variants as primitive ops

[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -271,12 +271,12 @@
 
 - func: view_as_real(Tensor(a) self) -> Tensor(a)
-  variants: function
+  variants: function, method
```
cc @anjali411 there were some good reasons not to have these as methods?

anjali411 replied:
There wasn't a specific reason why we didn't wanna add them as methods. I think it was more like that they were added as utility functions to help people migrate to the new complex types and should be migrated at some point. 

@bdhirsh is there a reason why you want to make `view_as_real` and `view_as_complex` available as methods?

bdhirsh replied:
Yeah I should have called this out more explicitly.

It simplifies the codegen a little, because those are currently the only two view ops that aren't methods.


Let me know if you'd rather not expand the api surface, and I can update the codegen

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3847,6 +3847,27 @@
+  device_check: NoCheck
+  device_guard: False
+  dispatch:
+    CPU, CUDA: slice_scatter
```
Why limit these to these two devices? Why not a CompositeAutogradExplicit?

bdhirsh replied:
So, I don't think we want other backends like xla/vulkan to use this implementation for two reasons:
- `slice_scatter` is used to implement the functionalization pass: when called inside of the pass, it's currently a primitive w.r.t the functionalization pass. Its implementation happens to use views and mutations, but that should be an implementation detail that doesn't leak into the pass, since the pass **uses** slice_scatter to remove aliasing in the first place.
- `slice_scatter` (for cpu/cuda at least) is implemented with `slice`, a view op. vulkan and xla (eventually) won't have actual implementations of these ops, since they'll rely on the functionalization pass to remove view ops (and replace them with view_copy calls).


But I feel like there's a mismatch between `slice_scatter` "the public API" and `slice_scatter` "in the implementation weeds of functionalization". If a user calls `slice_scatter` directly, it seems reasonable to make it composite - it should decompose into a `slice` and a `copy_` call, which will then get fed into the functionalization pass if you're running on xla. But when it's used inside of the functionalization pass, we don't want it to (and it currently doesn't) get recursively fed into the functionalization pass again (?).

Let me know if that makes sense (@ezyang). It seems to me like the options are (in order of implementation ease):
- Keep as a public API, only implemented for cpu/cuda (the code deserves a comment if we do this)
- Add two versions of the op: one thats public and composite, and one that's private and only used by the functionalization pass (this would make the logic uglier though, since we'd need to special-case these ops in the codegen)

albanD replied:
I think your point is valid and indeed backends like xla/vulkan won't be able to use this implementation and will have to specify their own.
But I don't think that's a reason to make all other backends (mkl/mlc/etc) have to implement it as well while this default implementation would work.

Just to be clear, I suggested a composite Explicit implementation, not implicit. So it will behave exactly the same as it does right now, just that other backends will also be able to use that by default if they support view/inplace!

bdhirsh replied:
Ok, I think that's fair.

The main thing I'm worried about is error detection: a backend opting into the functionalization pass, and getting silent failures because they forgot to register their own kernels for `slice_scatter` et.al.

We can probably worry about that for when I actually try to get vulkan/xla working with the pass though. At the very least, there are only two backends that we know want to use the pass. To harden it for future backends, maybe we'll want separate private API's for these functions in the future.

albanD commented on test/test_autograd.py:
```diff
@@ -2020,6 +2020,36 @@ def func(x):
         gradcheck(func, [x])
         gradgradcheck(func, [x])
 
+    def test_select_scatter(self):
```
Why not just an opinfo?

ezyang replied:
it seems the answer is, because the other view functions don't have op infos either

albanD replied:
They do. I did 2 search and found view and narrow. So I guess others are there too.

zou3519 replied:
We should add these as OpInfos so that someone else doesn't need to come along and do that in the future

bdhirsh replied:
Time to brush up on OpInfo authoring :)

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3847,6 +3847,27 @@
   dispatch:
     CompositeExplicitAutograd: slice
 
+- func: slice_scatter(Tensor self, Tensor mutated_view, int dim=0, int? start=None, int? end=None, int step=1) -> Tensor
```
The arguments name `mutated_view` doesn't look right.
All our scatter ops use `src` for the Tensor being scattered.

ezyang replied:
Well, they're not terribly consistent about it; `masked_scatter` takes `source` and `scatter` takes `src` except when it's a non-Tensor; then it's `value`. Agreed that a name like source is probably better though.

By the way, if we're talking about conventions, most of the pre-existing scatter functions take source at the very end, rather than the beginning. But I like Brian's new convention better as it handles optional arguments in the original view function better (as is the case in this example).

albanD replied:
Yes.
Also the name `mutated_view` is very specific to the way we want to use it in the context of the functionalization. But doesn't really make sense otherwise.

bdhirsh replied:
I'll rename to `mutated_view` to `src`

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1234,6 +1234,21 @@
   result: auto_linear
 
+- name: slice_scatter(Tensor self, Tensor mutated_view, int dim=0, int? start=None, int? end=None, int step=1) -> Tensor
+  self: grad * slice_scatter(at::ones(self.sizes(), self.options()), at::zeros(mutated_view.sizes(), mutated_view.options()), dim, start, end, step)
```
Why not just slice_scatter zeros into grad?

ezyang replied:
cc @zou3519 on use of `self.options()` here which seems questionable

zou3519 replied:
The use of self.options() here is probably OK. We don't have an invariant that "if `self` is a BatchedTensor, then its gradient should also be a BatchedTensor"; if it's necessary the gradient will get expanded later down the line

bdhirsh replied:
slice_scatter directly into grad makes sense, thanks

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1234,6 +1234,21 @@
 
+- name: slice_scatter(Tensor self, Tensor mutated_view, int dim=0, int? start=None, int? end=None, int step=1) -> Tensor
+  self: grad * slice_scatter(at::ones(self.sizes(), self.options()), at::zeros(mutated_view.sizes(), mutated_view.options()), dim, start, end, step)
+  mutated_view: grad * slice_scatter(at::zeros(self.sizes(), self.options()), at::ones(mutated_view.sizes(), mutated_view.options()), dim, start, end, step)
```
That is not the right shape?
This should just be a slice no?

ezyang replied:
How come gradcheck didn't catch this?

albanD replied:
Because it looks like a broadcasted argument I think :'(
And so the engine will reduce the gradient to match the shape of `mutated_view`...

bdhirsh replied:
Just to confirm - I think this should just be a tensor of ones, of the size of `mutated_view`.

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1234,6 +1234,21 @@
+  mutated_view: grad * slice_scatter(at::zeros(self.sizes(), self.options()), at::ones(mutated_view.sizes(), mutated_view.options()), dim, start, end, step)
+  result: auto_linear
+
+- name: select_scatter(Tensor self, Tensor mutated_view, int dim, int index) -> Tensor
```
same formula comments as above.

ezyang commented on test/test_view_ops.py:
```diff
@@ -397,6 +397,9 @@ def test_diagonal_view(self, device) -> None:
         self.assertEqual(t[0, 0, 1], v[0, 0])
 
+        inverse_t = t.diagonal_scatter(v, offset=1, dim1=1, dim2=2)
+        self.assertEqual(inverse_t, t)
```
This test doesn't seem to do what you want it to; as `v[0, 0] = 0` mutated the originating tensor `t`, so even if `diagonal_scatter` returned `t` unchanged, this test would pass anyway.

bdhirsh replied:
I'm going to kill this anyway in favor of the OpInfo tests

zou3519 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2456,4 +2456,21 @@ std::vector<Tensor> unflatten_dense_tensors(const Tensor& flat, TensorList tenso
 }
 
-}} // at::native
+at::Tensor slice_scatter(const at::Tensor& base, const at::Tensor& mutated_view, int64_t dim, c10::optional<int64_t> start, c10::optional<int64_t> end, int64_t step) {
+    auto output = base.clone();
+    output.slice(dim, start, end, step).copy_(mutated_view);
+    return output;
+}
+at::Tensor select_scatter(const at::Tensor& base, const at::Tensor& mutated_view, int64_t dim, int64_t index) {
+    auto output = base.clone();
+    output.select(dim, index).copy_(mutated_view);
+    return output;
+}
+at::Tensor diagonal_scatter(const at::Tensor& base, const at::Tensor& mutated_view, int64_t offset, int64_t dim1, int64_t dim2) {
+    auto output = base.clone();
+    output.diagonal(offset, dim1, dim2).copy_(mutated_view);
+    return output;
+}
```
Are we exposing these to python? (We probably should, if users want to write code without mutations). If so we should add docstrings for these

bdhirsh replied:
This PR currently will expose them to python. You're right, I should add doc strings. Is the right place for them in `torch/_torch_docs.py`? (I'll put them near their corresponding view docstrings)

zou3519 replied:
Yes, we want entries in `torch/_torch_docs.py`, and also entries in `torch/_tensor_docs.py` that link to the ones in `torch/_torch_docs.py`

For example, [tensor.abs documentation](https://pytorch.org/docs/stable/generated/torch.Tensor.abs.html?highlight=tensor%20abs#torch.Tensor.abs) links to [torch.abs docs](https://pytorch.org/docs/stable/generated/torch.abs.html#torch.abs)

bdhirsh replied:
Thanks. Although, I see that `slice` is currently not exposed to python (it's hardcoded in a [list in the codegen](https://github.com/pytorch/pytorch/blob/2853c7da2243611c6fbd8b78b93c0cd34f0f1bc1/tools/autograd/gen_python_functions.py#L85)). I'm thinking of doing the same for `slice_scatter` to avoid confusion.

zou3519 replied:
Aha. The reason why slice is not exposed to python is that it's literally just indexing with slices (e.g. `x[0:2]`). We might still want to expose a slice_scatter though for people who want to write code without mutations

bdhirsh replied:
Will do.

I also see that diagonal has both function/method docs, but slice is missing function docs; I figure I'll add them in for consistent (basically copying the method doc into the functions file, and adding a reference from [method docs] -> [function docs], which is what `diagonal` currently does)

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

ezyang commented on tools/autograd/derivatives.yaml:
```diff
@@ -1250,6 +1250,20 @@
   grad_output: grad.slice(dim, start, end, step)
+
+- name: slice_scatter(Tensor self, Tensor src, int dim=0, int? start=None, int? end=None, int step=1) -> Tensor
+  self: slice_scatter(grad, at::zeros(src.sizes(), src.options()), dim, start, end, step)
```
cc @zou3519, src.options police! @albanD I wonder if we could get away with using the mutable version of this in the formula here, saves the second copy

albanD replied:
I would actually expect that src would broadcast here and so you can do `at::zeros({}, ...)` to avoid this problem.

Also it is *never* allowed to modify `grad` inplace!

zou3519 replied:
This callsite isn't a problem for vmap because slice_scatter is an out-of-place function.

But could we rewrite `at::zeros(src.sizes(), src.options())` as `torch::zeros_like(srcs)`? That's pretty much what zeros_like does and it has the added benefit of being less verbose

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)




[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)


Differential Revision: [D31942092](https://our.internmc.facebook.com/intern/diff/D31942092)

[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)


Differential Revision: [D31942092](https://our.internmc.facebook.com/intern/diff/D31942092)

[ghstack-poisoned]

bdhirsh committed:
Update on "add slice/select/diagonal_scatter variants as primitive ops"

The functionalization pass needs `{view}_scatter` versions of the slice/select/diagonal ops in order to correctly propagate mutations from a view to its base. On top of that, the implementations need to be primitive w.r.t. autograd, because they look something like `...slice().copy_()`, and the functionalization pass can't use views + mutations inside of it's own alias-removal machinery!

I added some basic tests that I tried to base off of existing tests for views (particularly around testing the derivative formulas), but I'm wondering if I should add something more comprehensive.

Also, as_strided fits into this category - the functionalization pass will need an `as_strided_scatter` op that's primitive w.r.t. autograd. I didn't add it for now, because it'll involve duplicating a bunch of logic from the current `as_strided_backward()` function, and also writing a derivative formula that I wasn't sure how to write :)


Differential Revision: [D31942092](https://our.internmc.facebook.com/intern/diff/D31942092)

[ghstack-poisoned]
